packages/adapters/src/utilities/referencedMetadataProvider.ts

```diff
@@ -0,0 +1,170 @@
+import { metaData, Enums, type Types } from '@cornerstonejs/core';
+import dcmjs from 'dcmjs';
+
+const { DicomMetaDictionary } = dcmjs.data;
+const { MetadataModules } = Enums;
+
+export const STUDY_MODULES = [
+  MetadataModules.GENERAL_STUDY,
+  MetadataModules.PATIENT_STUDY,
+  MetadataModules.PATIENT,
+];
+
+export const SERIES_MODULES = [MetadataModules.GENERAL_SERIES];
+
+export const IMAGE_MODULES = [
+  MetadataModules.GENERAL_IMAGE,
+  MetadataModules.IMAGE_PLANE,
+  MetadataModules.CINE,
+  MetadataModules.VOI_LUT,
+  MetadataModules.MODALITY_LUT,
+  MetadataModules.SOP_COMMON,
+];
+
+/**
+ * Contains a metadata provider which knows how to generate different types of
+ * referenced metadata for things like creation a reference to an existing
+ * image, creating a new DICOM object entirely etc.
+ *
+ * There are also specific study/series/instance level attribute getters which
+ * can be used to get the attributes at a given level in the Normalized format
+ * instead of the lowerCamelCase format.  This assists in creating new DICOM
+ * objects.
+ *
+ * For example, this module provides a `ImageSopInstanceReference` implementation
+ * based on 'sopCommonModule' and 'frameModule' which references an image correctly
+ */
+
+export const metadataProvider = {
+  get: function (type: string, imageId: string, options) {
+    return metadataProvider[type]?.(imageId, options);
+  },
+
+  /**
+   * Returns an image sop instance reference for the given image, based on the
+   * frame module metadata.
+   * Note: UpperCamelCase for normalized response.
+   */
+  [MetadataModules.IMAGE_SOP_INSTANCE_REFERENCE]: function (imageId: string) {
+    const frameModule = metaData.get(MetadataModules.FRAME_MODULE, imageId);
+    const { sopClassUID, sopInstanceUID, frameNumber, numberOfFrames } =
+      frameModule;
+    if (numberOfFrames > 1) {
+      return {
+        ReferencedSOPClassUID: sopClassUID,
+        ReferencedSOPInstanceUID: sopInstanceUID,
+        ReferencedFrameNumber: frameNumber,
+      };
+    }
```

Comment:
Why can't we add these metadata module providers to wadors and wadouri? Having them inside feels a bit odd.

Reply:
This is the plan for how to fix the per-frame metadata modules and the calibration/spacing information as well - I will add another metadata module provider that knows how to deal with those types of operations as a combined provider.  Then, for instance normalized data it will just work without having to have the combine instance handling that breaks so many things in CS3D.